airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/state/CursorStateMessageProducer.java

```diff
@@ -0,0 +1,144 @@
+
+  private boolean cursorOutOrOrderDetected = false;
+
+  public CursorStateMessageProducer(final StateManager stateManager,
+                                    final String initialCursor) {
+    this.stateManager = stateManager;
+    this.initialCursor = initialCursor;
+    this.currentMaxCursor = initialCursor;
+  }
+
+  /**
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(final ConfiguredAirbyteStream stream) {
+    // At this stage intermediate state message should never be null; otherwise it would have been
+    // blocked by shouldEmitStateMessage check.
+    final AirbyteStateMessage message = intermediateStateMessage;
+    intermediateStateMessage = null;
+    if (cursorOutOrOrderDetected) {
+      LOGGER.warn("Intermediate state emission feature requires records to be processed in order according to the cursor value. Otherwise, "
+          + "data loss can occur.");
+    }
+    return message;
+  }
+
+  /**
+   * Note: We do not try to catch exception here. If error/exception happens, we should fail the sync,
+   * and since we have saved state message before, we should be able to resume it in next sync if we
+   * have fixed the underlying issue, of if the issue is transient.
+   *
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(final ConfiguredAirbyteStream stream, AirbyteMessage message) {
+    totalRecordCount++;
+    final String cursorField = IncrementalUtils.getCursorField(stream);
+    if (message.getRecord().getData().hasNonNull(cursorField)) {
+      final String cursorCandidate = getCursorCandidate(cursorField, message);
+      final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(stream,
+          cursorField);
+      final int cursorComparison = IncrementalUtils.compareCursors(currentMaxCursor, cursorCandidate, cursorType);
+      if (cursorComparison < 0) {
+        // Update the current max cursor only when current max cursor < cursor candidate from the message
+        if (!Objects.equals(currentMaxCursor, initialCursor)) {
+          // Only create an intermediate state when it is not the first record.
+          intermediateStateMessage = createStateMessage(stream, totalRecordCount);
+        }
+        currentMaxCursor = cursorCandidate;
+        currentMaxCursorRecordCount = 1L;
+      } else if (cursorComparison == 0) {
+        currentMaxCursorRecordCount++;
+      } else if (cursorComparison > 0) {
+        cursorOutOrOrderDetected = true;
+      }
+    }
+    return message;
+
+  }
+
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(final ConfiguredAirbyteStream stream) {
+    return createStateMessage(stream, totalRecordCount);
+  }
+
+  /**
+   * Only sends out state message when there is a state message to be sent out.
+   */
+  @Override
+  public boolean shouldEmitStateMessage(final ConfiguredAirbyteStream stream) {
```

Comment:
nit : One thing I noticed is that both `SourceStateIterator` and `SourceStateMessageProducer` have a `shouldEmitStateMessage` function. Do you think there is a good way to consolidate them? 

Maybe not in this PR, but I'm trying to remember why we have 2 different interfaces in the first place

Reply:
It's more of an additional check - in the shared SourceStateIterator we check how often to send (numOfRecords and time), and in the producer it applies additional check (such as there is no state message to send, so even if we hit numOfRecords we still do not send out state message)